tensorflow/core/kernels/training_ops.cc

```diff
@@ -418,6 +418,81 @@ struct ApplyPowerSign<CPUDevice, T> {
   }
 };
 
+template <typename T>
+struct ApplyIRpropPlus<CPUDevice, T> {
+  void operator()(const CPUDevice& d, typename TTypes<T>::Flat var,
+                  typename TTypes<T>::Flat old_grad,
+                  typename TTypes<T>::Flat delta_update,
+                  typename TTypes<T>::ConstScalar eta_minus,
+                  typename TTypes<T>::ConstScalar eta_plus,
+                  typename TTypes<T>::ConstScalar delta_min,
+                  typename TTypes<T>::ConstScalar delta_max,
+                  typename TTypes<T>::ConstScalar error,
+                  typename TTypes<T>::ConstScalar old_error,
+                  typename TTypes<T>::ConstFlat grad) {
+    auto sign_t = (old_grad * grad).sign();
+    // change to ColMaj to perform operations
+    Eigen::Tensor<T, 1> grad_sign = sign_t.swap_layout();
+
+    for (int i = 0; i < var.size(); i++) {
+      auto sign = grad_sign(i);
+
+      if (sign < T(0)) {
+        // revert deltaW(t-1) first then update delta
+        if (error() > old_error())
+          // compute deltaW and update the weight
+          var(i) -= delta_update(i) * (-sgn(old_grad(i)));
```

Comment:
I think this should be var.device(i) -= ... (same below)

Reply:
the code is correct (var is a Flat tensor and each coefficient is updated individually according to the sign of the gradients), the implementation for cpu is not multi threaded.